src/Microsoft.ML.StandardLearners/Standard/SdcaStatic.cs

```diff
@@ -54,17 +56,16 @@ public static partial class RegressionTrainers
             Contracts.CheckParam(!(l1Threshold < 0), nameof(l1Threshold), "Must not be negative, if specified.");
             Contracts.CheckParam(!(maxIterations < 1), nameof(maxIterations), "Must be positive if specified");
             Contracts.CheckValueOrNull(loss);
+            Contracts.CheckValueOrNull(advancedSettings);
             Contracts.CheckValueOrNull(onFit);
 
-            var args = new SdcaRegressionTrainer.Arguments()
-            {
-                L2Const = l2Const,
-                L1Threshold = l1Threshold,
-                MaxIterations = maxIterations
-            };
+            var args = new SdcaRegressionTrainer.Arguments();
+            advancedSettings?.Invoke(args);
+            args.L2Const = l2Const;
+            args.L1Threshold = l1Threshold;
+            args.MaxIterations = maxIterations;
```

Comment:
for some of them,  i checked whether the users has set l2, l1 and maxIterations in the advancedSettings, rather then pass the arguments directly to the method, and warned about overriding those with the defaults. Feel like we owe that that, since we are providing two places to set them.

Reply:
In any case, the correct logic is to set the params first, then invoke the delegate

---
In reply to: [223046571](https://github.com/dotnet/machinelearning/pull/1160#discussion_r223046571) [](ancestors = 223046571)

Reply:
I'll stick to Pete approach if you don't mind

---
In reply to: [222897010](https://github.com/dotnet/machinelearning/pull/1160#discussion_r222897010) [](ancestors = 222897010)